cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java

```diff
@@ -0,0 +1,188 @@
+    String runtimeJobClassName = args[0];
+
+    ClassLoader cl = DataprocJobMain.class.getClassLoader();
+    if (!(cl instanceof URLClassLoader)) {
+      throw new RuntimeException("Classloader is expected to be an instance of URLClassLoader");
+    }
+
+    // get classpath
+    Deque<URL> queue = getClasspath((URLClassLoader) cl);
+
+    // create new URL classloader with provided classpath
+    try (URLClassLoader newCL = new URLClassLoader(queue.toArray(new URL[0]), cl.getParent())) {
+      Thread.currentThread().setContextClassLoader(newCL);
+
+      // load environment provider class and create instance of it
+      String runtimeProviderClassName = DataprocRuntimeEnvProvider.class.getName();
+      Class<?> envProviderClass = newCL.loadClass(runtimeProviderClassName);
+      Object newEnvProviderInstance = envProviderClass.newInstance();
+
+      try {
+        // call initialize() method on envProviderClass
+        Method initializeMethod = envProviderClass.getMethod("initialize");
+        System.out.println("Invoking initialize() on " + runtimeProviderClassName);
+        initializeMethod.invoke(newEnvProviderInstance);
+
+        // call getRuntimeJobEnvironment() method on envProviderClass
+        Method getRuntimeEnvMethod = envProviderClass.getMethod("getRuntimeJobEnvironment");
+        System.out.println("Invoking getRuntimeJobEnvironment() " + runtimeProviderClassName);
+        Object runtimeEnv = getRuntimeEnvMethod.invoke(newEnvProviderInstance);
+
+        // call run() method on runtimeJobClass
+        Class<?> runEnvCls = newCL.loadClass(RuntimeJobEnvironment.class.getName());
+        Class<?> runner = newCL.loadClass(runtimeJobClassName);
+        Method method = runner.getMethod("run", runEnvCls);
+        System.out.println("Invoking run() on " + runtimeJobClassName);
+        method.invoke(runner.newInstance(), runtimeEnv);
+
+      } finally {
+        // call close() method on envProviderClass
+        Method closeMethod = envProviderClass.getMethod("close");
+        System.out.println("Invoking close() " + runtimeProviderClassName);
+        closeMethod.invoke(newEnvProviderInstance);
+      }
+
+      System.out.println("Runtime job completed.");
+    }
+  }
+
+  private static Deque<URL> getClasspath(URLClassLoader cl) throws IOException {
+    File appJarDir = new File(Constants.Files.APPLICATION_JAR);
+    File twillJarDir = new File(Constants.Files.TWILL_JAR);
+    File resourceJarDir = new File(Constants.Files.RESOURCES_JAR);
+    URL[] classpath = createClasspath(appJarDir, twillJarDir, resourceJarDir);
+    URL[] urls = cl.getURLs();
+    List<URL> urlList = new ArrayList<>(Arrays.asList(urls));
+    Deque<URL> queue = new LinkedList<>(urlList);
+    for (URL url : classpath) {
+      if (url.toString().endsWith(".jar")) {
+        addAllDeps(url, queue);
+        queue.addFirst(url);
+      } else {
+        queue.addLast(url);
+      }
+    }
+    System.out.println("Classpath URLs: " + queue);
```

Comment:
Use LOG instead.

Reply:
I followed the pattern similar to RemoteLauncher#main(). But I will change it to use logger.